benchmarks/canny_edge_detector/functional_c#/Program.cs

```diff
@@ -0,0 +1,170 @@
+            })
+            .ToImmutableArray();
+        }
+
+        static readonly ImmutableArray<(int x, int y)> range = Enumerable.Range(-1, 3).SelectMany(x => Enumerable.Range(-1, 3).Select(y => (x, y))).ToImmutableArray();
+        private static bool strongNeighbour(ImmutableArray<(int x, int y, int w)> pic, int width, int height, int x, int y)
+        {
+            return range
+                .Any(f => {
+                    var posX = x + f.x;
+                    var posY = y + f.y;
+                    return !((f.x == 1 && f.y == 1) || posX <= 0 || posX >= width - 1 || posY <= 0 || posY >= height - 1) 
+                            && pic[posY * width + posX].w == 255;
+                });
+        }
+
+        private static ImmutableArray<(int, int, int)> doubleThreshold(ImmutableArray<(int x, int y, int w)> pic)
+        {
+            var max = pic.Max(p => p.w);
+            var high = max * 0.09;
+            var low = high * 0.5;
+            return pic.Select(p => (p.x, p.y, p.w <= low ? 0 : (p.w < high ? weak : 255)))
+                    .ToImmutableArray();
+        }
+
+        private static ImmutableArray<(int, int, int)> nonMaxSupression((ImmutableArray<(int x, int y, int)> pic, ImmutableArray<(int, int, int)> direction) input)
+        {
+            var (pic, direction) = input;
+            var temp = pic.Last();
+            var width = temp.x + 1;
+            var height = temp.y + 1;
+
+            return pic.Zip(direction, (g, d) => (g.x, g.y, maxSupressionOne(width, height, pic, g.x, g.y, g.Item3, d.Item3)))
+                    .ToImmutableArray();
+        }
+
+        private static int maxSupressionOne(int width, int height, ImmutableArray<(int x, int y, int)> pic, int x, int y, int w1, int w2)
+        {
+            if(x == 0 || x == width - 1 || y == 0 || y == height - 1)
+                return w1;
+
+            var tq = w2 % 4;
+            if((tq == 0 && (w1 <= pic[x+(y-1)*width].Item3   || w1 <= pic[x+(y+1)*width].Item3)) //0 is E-W (horizontal)
+            ||(tq == 1 && (w1 <= pic[x-1+(y+1)*width].Item3 || w1 <= pic[x+1+(y-1)*width].Item3)) //1 is NE-SW
+            ||(tq == 2 && (w1 <= pic[x-1+y*width].Item3     || w1 <= pic[x+1+y*width].Item3)) //2 is N-S (vertical)
+            ||(tq == 3 && (w1 <= pic[x-1+(y-1)*width].Item3 || w1 <= pic[x+1+(y+1)*width].Item3))) //#3 is NW-SE
+                return 0;
+            else
+                return w1;
+        }
+
+        private static (ImmutableArray<(int, int, int)> gradient, ImmutableArray<(int, int, int)> direction) intensityGradients(ImmutableArray<(int x, int y, int colour)> pic)
+        {
+            var hori = convolve(pic, kernelHori);
+            var vert = convolve(pic, kernelVert);
+
+            var gradiant = hori.Zip(vert, (h, v) => (h.x, h.y, hyp(h.Item3, v.Item3))).ToImmutableArray();
+            var direction = hori.Zip(vert, (h, v) => (h.x, h.y, arctan(h.Item3, v.Item3))).ToImmutableArray();

```

Comment:
Kan man slå de her sammen, så vi kun går gennem listen en enkelt gang i stedet for to?